packages/server/src/create-instance/extract-critical2.js

```diff
@@ -0,0 +1,38 @@
+// @flow
+import type { EmotionCache } from '@emotion/utils'
+
+const createExtractCritical2 = (cache: EmotionCache) => (html: string) => {
+  // parse out ids from html
+  // reconstruct css/rules/cache to pass
+  let RGX = new RegExp(`${cache.key}-([a-zA-Z0-9-_]+)`, 'gm')
+
+  let o = { html, ids: [], css: '', globalCss: [] }
+  let match
+  let ids = {}
+  while ((match = RGX.exec(html)) !== null) {
+    // $FlowFixMe
+    if (ids[match[1]] === undefined) {
+      // $FlowFixMe
+      ids[match[1]] = true
+    }
+  }
+
+  o.ids = Object.keys(cache.inserted).filter(id => {
+    if (
+      (ids[id] !== undefined ||
+        cache.registered[`${cache.key}-${id}`] === undefined) &&
+      cache.inserted[id] !== true
+    ) {
+      if (cache.registered[`${cache.key}-${id}`]) {
+        o.css += cache.inserted[id]
+      } else {
+        o.globalCss.push(cache.inserted[id])
```

Comment:
Since we can end up with multiple `<Global/>` or `injectGlobal` calls, I think we need to return a new object for each global style (since we should construct a style element per each global style to isolate them and make them flushable independently).

I imagine that `extractCritical2` would return `Array<{ key: string, ids: string[], css: string }>`. Not completely sure yet if we need both key and ids or if this could be simplified without diving more into the code. By adding tests for independent flushing this should become more apparent what is needed here. We might not need to differentiate between regular and global "chunks" (elements of the proposed return type) since the returned info should be enough for the HTML to be generated but I might be wrong on this one.

Reply:
Ah I see that makes sense, I believe that for the regular `css` we should have an array of ids, but for the global items, it would theoretically be one id. Let me try to refactor and make sure that that would work. Will add test cases for independent flushing.

Reply:
Updated, let me know if this is closer to what you had in mind